File: Api-Gataway/src/main/java/it/uniroma3/domain/UserService.java
package it.uniroma3.domain;


import it.uniroma3.common.event.DomainEventPublisher;
import it.uniroma3.event.UserCreatedEvent;
import it.uniroma3.event.UserServiceChannel;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@Service
public class UserService implements IUserService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder bCryptPasswordEncoder;

    @Autowired
    private DomainEventPublisher domainEventPublisher;

    public User create(String username, String password, String firstname, String lastname){
        User user = new User(username, password);
        user.setEnabled(true);
        user.setPassword(bCryptPasswordEncoder.encode(user.getPassword()));
        Authority authorityUser=new Authority();
        authorityUser.setName(AuthorityName.ROLE_USER);
        List<Authority> authorities = new ArrayList<>();
        authorities.add(authorityUser);
        user.setAuthorities(authorities);
        userRepository.save(user);
        //genero l'evento di creazione di un nuovo utente
        UserCreatedEvent event = makeUserCreatedEvent(user.getId(), firstname, lastname);
        domainEventPublisher.publish(event, UserServiceChannel.userServiceChannel);
        return user;
    }

    public User createAmdmin(User user){
        return userRepository.save(user);
    }

    private UserCreatedEvent makeUserCreatedEvent(Long id, String firstname, String lastname){
        return new UserCreatedEvent(id,firstname,lastname);
    }

    @Override
    public User findByUsername(String username) {
        return userRepository.findByUsername(username).orElse(null);
    }
}

File: iam/src/main/kotlin/it/uniroma3/domain/service/UserService.kt
package it.uniroma3.domain.service


import it.uniroma3.config.WebSecurityUser
import it.uniroma3.domain.model.RoleName
import it.uniroma3.domain.model.User
import it.uniroma3.domain.repository.RoleRepository
import it.uniroma3.domain.repository.UserRepository
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service

@Service
class UserService (private val userRepository: UserRepository,
                   private val roleRepository: RoleRepository) : UserDetailsService {


    override fun loadUserByUsername(username: String): UserDetails {
        val findByUserName = userRepository.findByUsername(username)
        return WebSecurityUser(findByUserName)
    }

    fun createUser(username: String, password: String,
                   email : String, firstName : String, lastName: String) : User {
        val passwordEncr = BCryptPasswordEncoder().encode(password)
        var user = User(firstName, lastName, username, email, passwordEncr)
        user=userRepository.save(user)
        return user
    }

    fun toUser(user: User) : User {
        val role = roleRepository.findByName(RoleName.User.name)
        val newRoles = HashSet(user.roles)
        newRoles.add(role)
        user.roles=newRoles
        return userRepository.save(user)
    }

}

